### ISSUE TITLE
Droppping experimental_setReleaseTimeoutMs

### ISSUE THREAD

**Issue by stari4ek** [google/ExoPlayer#7405]
### [REQUIRED] Searched documentation and issues
source code itself

### [REQUIRED] Question
There was experimental feature to specify timeout for `release()`, which was dropped recently (https://github.com/google/ExoPlayer/commit/78c850a88579beb5ebf2c487a0bdd45a97d0c881)

I'm interested why it was dropped.
Here is my use-case:

According to crashlytics and google play console, there is issue, when ExoPlayer `release()` takes pretty long time: 5-15sec.

I'm using ExoPlayer to implement `TV Input Framework` integration, so ExoPlayer is used inside `TvContract` session.
Session lifetime is defined by system TV App (like _Live Channels_). There're multiple timeout guards, which crashes service, if processing specific message takes longer than threshold (5sec).

So, sometime when my service releases ExoPlayer or releases surface - ExoPlayer blocks main thread for 5sec+ and session is punished for it (as ANR by system, or by those guards).

According to the log - it happens when `TvContract` session is going to be released. Maybe cause device is going to sleep/shut down.

So, in this case - it was really useful to specify short timeout - to guarantee, that service (and it's process, cause there is unhandled exception) won't be killed by system or system wrapper around it.

**Comment by christosts** [comment]
This API was added to help us debug/resolve ANRs we see in the field. We wanted to see if ANRs are caused by ExoPlayer.release() and we added the timeout so these ANRs would now be Timeout exceptions. Note that this would not solve the underlying issue that something is blocking Exoplayer.release(). If ExoPlayer.release() is blocked and times out, it's highly likely that underlying resources (like a MediaCodec instance) are not released, therefore leading to undefined device behavior. 

The purpose of the API was to make UX somewhat better (raise custom error message instead of ANR) and help us measure the impact of ANRs possibly caused by Exoplayer.release(). However, when we tried this, we did not observe/measure drop in ANRs, so the original hypothesis was not verified; the ANRs we were investigating are not caused by ExoPlayer.release().

Do you have data saying that ANRs are indeed dropped when you set a timeout on release()?

**Comment by stari4ek** [comment]
This functionality was adopted recently. So I need at least 1-2 weeks to collect enough data to compare.

BTW, this is how it looks like:

```
820 | Tue May 19 2020 20:03:44 GMT+0300 | main|T] Session-0: onSetSurface: [null]
821 | Tue May 19 2020 20:03:51 GMT+0300 | main|D] EventLogger: surfaceSize [eventTime=2338.67, mediaPos=59.13, window=0, period=0, 0, 0]
```

7sec to release surface

or even more:
```
08:29:02.398 AM main|T] Session-0: onSetSurface: [null]
08:29:47.094 AM main|D] EventLogger: surfaceSize [eventTime=319.21, mediaPos=0.00, window=0, period=0, 0, 0]

Fatal Exception: java.lang.RuntimeException: Too much time to handle a request. (type=3, 44754ms > 5000ms).
       at android.media.tv.ITvInputSessionWrapper.executeMessage(ITvInputSessionWrapper.java:241)
       at com.android.internal.os.HandlerCaller$MyHandler.handleMessage(HandlerCaller.java:37)
```

Theoretically I can re-schedule surface release to next frame in main thread, to get rid of those guard, but I will get ANR for service anyway, since `release()` is processed in main thread.
ANR:
```
  at java.lang.Object.wait (Object.java)
- waiting on <0x05eb3c77> (a e.e.a.b.y0)
  at com.google.android.exoplayer2.PlayerMessage.blockUntilDelivered (PlayerMessage.java:290)
  at com.google.android.exoplayer2.SimpleExoPlayer.setVideoSurfaceInternal (SimpleExoPlayer.java:1849)
- locked <0x05eb3c77> (a e.e.a.b.y0)
  at com.google.android.exoplayer2.SimpleExoPlayer.clearVideoSurface (SimpleExoPlayer.java:534)
  at by.stari4ek.iptv4atv.player.PlayerController.setSurface (PlayerController.java:431)
  at by.stari4ek.iptv4atv.tvinput.service.SessionPlayerMixin.onSetSurface (SessionPlayerMixin.java:141)
  at by.stari4ek.iptv4atv.tvinput.service.SessionVerboseMixin.onSetSurface (SessionVerboseMixin.java:61)
  at android.media.tv.TvInputService$Session.setSurface (TvInputService.java:1158)
  at android.media.tv.ITvInputSessionWrapper.executeMessage (ITvInputSessionWrapper.java:132)
```

**Comment by christosts** [comment]
The API is back in commit [1154e80](https://github.com/google/ExoPlayer/commit/1154e8098ab079f5faf9c2cd1decc2339468c118). We'd appreciate if you provide any findings once you collect data. Thanks!

**Comment by stari4ek** [comment]
Great. Thank you,

I've adopted it for `ExoPlayer.release()`. So it clearly minimizes specific repro path:

![image](https://user-images.githubusercontent.com/327778/82748522-255c5d00-9dab-11ea-90e2-d3c50951bd78.png)

```
"main" prio=5 tid=1 Waiting
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x75021f50 self=0xf013d000
  | sysTid=11764 nice=0 cgrp=default sched=0/0 handle=0xf44b4494
  | state=S schedstat=( 2671655552 3369135715 5810 ) utm=203 stm=63 core=1 HZ=100
  | stack=0xff020000-0xff022000 stackSize=8MB
  | held mutexes=
  at java.lang.Object.wait (Object.java)
- waiting on <0x0ae5bfbb> (a e.e.a.b.c0)
  at com.google.android.exoplayer2.ExoPlayerImplInternal.waitUntilReleased (ExoPlayerImplInternal.java:519)
- locked <0x0ae5bfbb> (a e.e.a.b.c0)
  at com.google.android.exoplayer2.ExoPlayerImplInternal.release (ExoPlayerImplInternal.java:332)
- locked <0x0ae5bfbb> (a e.e.a.b.c0)
  at com.google.android.exoplayer2.ExoPlayerImpl.release (ExoPlayerImpl.java:681)
  at com.google.android.exoplayer2.SimpleExoPlayer.release (SimpleExoPlayer.java:1571)
  at by.stari4ek.iptv4atv.player.PlayerController.releasePlayer (PlayerController.java:767)
  at by.stari4ek.iptv4atv.player.PlayerController.release (PlayerController.java:264)
  at by.stari4ek.iptv4atv.tvinput.service.SessionPlayerMixin.releasePlayer (SessionPlayerMixin.java:208)
  at by.stari4ek.iptv4atv.tvinput.service.SessionPlayerMixin.onRelease (SessionPlayerMixin.java:148)
  at by.stari4ek.iptv4atv.tvinput.service.SessionOverlayMixin.onRelease (SessionOverlayMixin.java:104)
  at by.stari4ek.iptv4atv.tvinput.service.TvirlSession.onRelease (TvirlSession.java:141)
  at by.stari4ek.iptv4atv.tvinput.service.SessionVerboseMixin.onRelease (SessionVerboseMixin.java:139)
  at android.media.tv.TvInputService$Session.release (TvInputService.java:1132)
  at android.media.tv.ITvInputSessionWrapper.executeMessage (ITvInputSessionWrapper.java:114)
  at com.android.internal.os.HandlerCaller$MyHandler.handleMessage (HandlerCaller.java:37)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:193)
  at android.app.ActivityThread.main (ActivityThread.java:6680)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:858)
```

But it's better to watch if there are any side effects for it.

Also I will check how it works for `clearVideoSurface()`. Overall ANR rate is x4 higher for it, since `setSurface(null)` is always called prior releasing `TvContract` session.

**Comment by stari4ek** [comment]
Ah. I had false assumption that there is same kind "timeout" functionality for `clearVideoSurface()`. But there is no one.

I have bunch of 10-60-sec timeouts there. Like posted above.
Here is another one:

```
Fatal Exception: java.lang.RuntimeException: Too much time to handle a request. (type=3, 40929ms > 5000ms).
       at android.media.tv.ITvInputSessionWrapper.executeMessage(ITvInputSessionWrapper.java:241)

04:11:52.314 PM main|T] Session-0: onSetSurface: [null]
04:12:33.235 PM main|D] EventLogger: surfaceSize [eventTime=71.21, mediaPos=5.28, window=0, period=0, 0, 0]
```

Is it better to create new issue?

**Comment by christosts** [comment]
Regarding the data you collected from timeouts on ExoPlayer.release(). Can you tell what happened *after* the error occurred? Our internal findings suggested that users did not watch more videos after the exception was thrown, probably users had to shutdown the app, or even the restart the phone maybe. So in terms of overall video playback time, we did not see any improvement. Therefore we removed the API.

It'd be good to have data that this API does improve things so that we decide to make it permanent.  

Regarding the `clearVideoSurface()`: it'd be good to have a separate issue so that the discussion focuses on the issue at hand, so feel free to raise a separate issue. Please bear in mind that usually these issues are device specific and there's high probability we will not be able to offer a solution.

**Comment by stari4ek** [comment]
In my case - "app" is service, which is bound to Android TV's "TV App" (Live Channels/Live TV/OEM variant) and works inside it.
Right now I re-use ExoPlayer instance during whole session lifetime, and release it only when session is released (and in rare cases of OOM during playback, but i do not set timeout when restarting due OOM).
So I'm expecting next possible variants after `setTimeout()` +`release()` are used:
1. device is going to shut down or sleep. this is very possible use case: "turn device off while watching TV".
2. TV App is going to be closed. New session will be created with new surface after.
3. Session is released for short period (another UI is in foreground, voice command has been issued, ...). New session with new surface will be created.

It's really hard to trace what will happen after, but I'm expecting that preparing new instance of ExoPlayer with new surface won't have any issues. Not sure about decoders (especially for 2nd/3rd cases).

ATM I do not track metrics for session lifetime or app instance lifetime (if there're multiple sessions spawned during app's lifetime), so do not have any insights here.

On the positive side - I do not see any regressions for ExoPlayer preparations, which are theoretically possible for cases 2-3, if decoders are left in "broken" state.

**Comment by christosts** [comment]
I'll ping again in about 2 weeks time to get some fresh data.

**Comment by tonihei** [comment]
> However, when we tried this, we did not observe/measure drop in ANRs, so the original hypothesis was not verified

It turned out our measurements did not test this change correctly and using a timeout does indeed help to prevent ANRs. We will include the timeout in our code as default.

**Comment by tonihei** [comment]
I'll close this issue because the main issue is tracked by #4352 which will be updated once the change is out.
